packages/itwinui-react/src/core/Menu/MenuItem.tsx

```diff
@@ -159,92 +206,123 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+        size={size}
+        active={isSelected}
+        disabled={disabled}
+        ref={useMergedRefs(
+          menuItemRef,
+          forwardedRef,
+          subMenuItems.length > 0 ? popover.refs.setReference : null,
+        )}
+        role={role}
+        tabIndex={disabled || role === 'presentation' ? undefined : -1}
+        aria-selected={isSelected}
+        aria-haspopup={subMenuItems.length > 0 ? 'true' : undefined}
+        aria-controls={subMenuItems.length > 0 ? submenuId : undefined}
+        aria-expanded={subMenuItems.length > 0 ? popover.open : undefined}
+        aria-disabled={disabled}
+        {...(subMenuItems.length === 0
+          ? { ...handlers, ...rest }
+          : popover.getReferenceProps({ ...handlers, ...rest }))}
+      >
+        {startIcon && (
+          <ListItem.Icon as='span' aria-hidden>
+            {startIcon}
+          </ListItem.Icon>
+        )}
+        <ListItem.Content>
+          <div>{children}</div>
+          {sublabel && <ListItem.Description>{sublabel}</ListItem.Description>}
+        </ListItem.Content>
+        {!endIcon && subMenuItems.length > 0 && (
+          <ListItem.Icon as='span' aria-hidden>
+            <SvgCaretRightSmall />
+          </ListItem.Icon>
+        )}
+        {endIcon && (
+          <ListItem.Icon as='span' aria-hidden>
+            {endIcon}
+          </ListItem.Icon>
+        )}
+
+        {subMenuItems.length > 0 && popover.open && (
+          <FloatingNode id={nodeId}>
+            <Portal>
+              <MenuItemContext.Provider
+                value={{
+                  ref: menuItemRef,
+                  setIsSubmenuVisible,
+                  setIsNestedSubmenuVisible,
+                }}
+              >
+                <Menu
+                  setFocus={focusOnSubmenu}
+                  ref={popover.refs.setFloating}
+                  {...popover.getFloatingProps({
+                    id: submenuId,
+                    onPointerMove: () => {
+                      // pointer might move into a nested submenu and set isSubmenuVisible to false,
+                      // so we need to flip it back to true when pointer re-enters this submenu.
+                      setIsSubmenuVisible(true);
+                    },
+                  })}
+                >
+                  {subMenuItems}
+                </Menu>
+              </MenuItemContext.Provider>
+            </Portal>
+          </FloatingNode>
+        )}
+      </ListItem>
+    </ConditionalFloatingTreeWrapper>
   );
 }) as PolymorphicForwardRefComponent<'div', MenuItemProps>;
+
+// ----------------------------------------------------------------------------
+
+export type TreeEvent = {
+  nodeId: string;
+  parentId: string | null;
+};
+
+/**
+ * Wraps `children` with `<FloatingTree>` if a `FloatingTree` context is not found.
+ *
+ * This is useful when `MenuItem`s are used outside of a `DropdownMenu`.
+ */
+const ConditionalFloatingTreeWrapper = (props: {
```

Comment:
I noticed that `MenuItem`s were used outside of `DropdownMenu` in the unit tests in `MenuItem.test.tsx`. So, I thought to have something like `ConditionalFloatingTreeWrapper` to add a `FloatingTree` when it does not exist.

Open to other ideas to implement this.

Reply:
> `MenuItem`s were used outside of `DropdownMenu` in the unit tests in `MenuItem.test.tsx`.

Why is this a problem? It should be fine.

Also we should almost never need to change our code for the purpose of making unit tests pass.

Reply:
> Why is this a problem? It should be fine.

We are currently relying on `FloatingTree` + `FloatingNode` to send events when a submenu opens. These events help us enforce only one submenu open at a time for each menu. Since tree events are not possible without a `FloatingTree` context as an ancestor, I added `ConditionalFloatingTreeWrapper`.

> Also we should almost never need to change our code for the purpose of making unit tests pass.

True, but I was trying to catch even non-unit test cases where `MenuItem`s could be outside the `DropdownMenu`.

If this case doesn't exist, then I can remove `ConditionalFloatingTreeWrapper` and update the unit tests accordingly.

Reply:
It might be better to test these interactions in a playwright test, with the whole component. Unit tests are a bad place to test such important, complex logic.

I don't think `MenuItem` should be used anywhere else. The only place I can think of is in the custom item renderers of `Select` and `ComboBox`, but those should use `ListItem` imo. We might want to look into showing a runtime warning to enforce this.

Reply:
> I don't think `MenuItem` should be used anywhere else.

For now, I removed `ConditionalFloatingTreeWrapper` since one more disadvantage of using it is that the top-level menu will not have one FloatingTree. Instead, each top-level menu's MenuItems will have a FloatingTree each. So, enforcing just one active submenu for the first level menu will not be possible with this approach.

> Or even deprecate `MenuItem` altogether and replace it with something like `DropdownMenu.Item`.

If we deprecate `MenuItem`, do we still need to support it until it is removed? If so, then the problem will exist for users that are using `MenuItem` instead of `DropdownMenu.Item`.

> The only place I can think of is in the custom item renderers of `Select` and `ComboBox`, but those should use `ListItem` imo. We might want to look into showing a runtime warning to enforce this. 

`Select` and `ComboBox` would not have submenu items. And the floating tree events are only for submenu management. So, even if the `<FloatingTree>` wrapper is not present, I don't think it would be a problem.

To be safer, should/can we add a `<FloatingTree>` inside `Select` and `ComboBox`, similar to how we're doing for `DropdownMenu`?

Reply:
> To be safer, should/can we add a `<FloatingTree>` inside `Select` and `ComboBox`, similar to how we're doing for `DropdownMenu`?

No, there's no need. I don't see how that's safer, since `ComboBox`/`Select` don't have submenus.

> If we deprecate `MenuItem`, do we still need to support it until it is removed? If so, then the problem will exist for users that are using `MenuItem` instead of `DropdownMenu.Item`.

Yes we need to support it, but that's not a difficult thing (I'm not sure what you mean by "the problem").

The point of a deprecation here is mainly to enforce correct usage. Currently if you use `MenuItem` inside `ComboBox`/`Select`, it gets incorrect semantics and needs to execute a bunch of unrelated code that gets thrown away anyway (because it is only relevant for `DropdownMenu`).

Reply:
> I'm not sure what you mean by "the problem"

By "the problem", I meant that even if we deprecate `MenuItem` inside`Select`/`ComboBox`, if a user still uses a `MenuItem` there, enforcing only one active submenu per menu will not be possible without wrapping the entire tree in `FloatingTree`. But I guess that since submenus should anyways not be there in `Select`s/`ComboBox`es, "this program" will not happen.